rfd/0153-resource-guidelines.md

```diff
@@ -0,0 +1,725 @@
+Legacy resources also had a `types.ResourceHeader` that used gogo magic to embed the type in the resource message. To
+get around this, the required fields from the header MUST be included in the message itself. A non-gogo clone does exist
+`teleport.header.v1.ResourceHeader`, however, to get the fields, embeded custom marshalling must be manually written.
+
+If a resource has associated secrets (password, private key, jwt, mfa device, etc.) they should be defined in a separate
+resource and stored in a separate key range in the backend. The traditional pattern of defining secrets inline and only
+returning them if a `with_sercrets` flag is provided causes a variety of problems and introduces opportunity for human
+error to accidentally include secrets when they should not have been. It would then be the responsibility of the caller
+to get both the base resource and the corresponding secret resource if required.
+
+There are many things to consider, and many ways to design a resource
+specification. To provide consistency and uniformity there are a few
+[standards](https://cloud.google.com/apis/design/standard_fields)
+and [design patterns](https://cloud.google.com/apis/design/design_patterns)
+that should be followed when possible. The most notable of the design patterns
+is
+[Bool vs. Enum vs. String](https://cloud.google.com/apis/design/design_patterns#bool_vs_enum_vs_string). There have been several occasions in the past where a particular field
+was not flexible enough which prevented behavior from being easily extended to
+support a new feature.
+
+All properties defined in the `spec` of a resource MUST only be modifed by the
+owner/creator of the resource. For example, if a resource is created via
+`tctl create`, then any fields within the `spec` MUST not be altered dynamically
+by the Teleport process. When Teleport automatically modifies the `spec` during
+runtime it causes drift between external IaC. If a resource has properties that
+are required to be modified dynamically by Teleport, a separate `status` field
+should be added to the resource to contain them. These fields will be ignored
+by IaC tools during their reconcilliation.
+
+### API
+
+All APIs should follow the conventions listed below that are largely based on
+the [Google API style guide](https://cloud.google.com/apis/design/standard_methods).
+
+#### Create
+
+The `Create` RPC takes a resource to be created and must also return the newly created resource so that any fields that
+are populated server side are provided to clients without requiring an additional call to `Get`.
+
+The request MUST fail and return a `trace.AlreadyExists` error if a matching resource is already present in the backend.
+
+```protobuf
+// Creates a new Foo resource in the backend.
+    rpc CreateFoo(CreateFooRequest) returns (Foo);
+
+message CreateFooRequest {
+  // The desired Foo to be created.
+  Foo foo = 1;
+}
+```
+
+#### Update
```

Comment:
There's some discussion of update/upsert over in https://github.com/gravitational/teleport/pull/35441#discussion_r1428614490, I will suggest we move that here.

Reply:
What about my point about the uselessness of patch-style operations under this paradigm (which pertains to your `FieldMask` comment below). As I understand it, in order to do an `UpdateFoo` with optimistic locking, you will need to have first done a `GetFoo` to get the object's revision for the call to `Update`. So then wouldn't that make the `FieldMask` useless, since you'll have the entire object on-hand anyhow?

Reply:
Got it, yeah that use case makes sense.

That brings up a question for me, though: do we need to implement a patch-style fieldmask operation in each supported `Backend` individually. Or is there some existing abstraction that takes care of that for us? I don't know that I've ever actually added a new resource to Teleport, so I'm not quite sure what that process entails generally, nor for a resource that supports FieldMask specifically.

Reply:
Thanks, that's similar although I think I'm also asking a further question -- as I understand it we support various backends like sqlite, etcd, dynamodb, etc. In adding a new resource, would we also need to understand how patching works in each of those and extend them?

Regardless of the answer, since this is likely to be used as a reference doc it might be worthwhile including a link to somewhere in the codebase where field masks were done well.

Reply:
The new BotService (https://github.com/gravitational/teleport/pull/35441) also makes use of field_masks for patching.